### PR TITLE
Refactoring the Marlowe.Execution module(s).

### PR DESCRIPTION
**Pull request by merivale**
While trying to get my head around what's going on in `Marlowe.Execution` and `Contract`, I felt the urge to do a bit of refactoring. Firstly, separating `Marlowe.Execution` out into `Types`, `Lenses`, and `State` submodules. Then renaming some things and moving some stuff around a little bit. Then adding some more comments.

This is an entirely optional PR. But see if you like it.

### CODE REVIEW

File: marlowe-dashboard-client/src/Contract/State.purs [hunk 1]
```diff
@@ -294,9 +297,9 @@ transactionsToStep { participants } { txInput, state } =
 timeoutToStep :: State -> Slot -> PreviousStep
 timeoutToStep { participants, executionState } slot =
   let
-    currentContractState = executionState ^. _currentState
+    currentMarloweState = executionState ^. _currentMarloweState
```
Comment (hrajchert) [inline marlowe-dashboard-client/src/Contract/State.purs:300]: same discussion as above... maybe something in the lines of `currentSemanticState`

File: marlowe-dashboard-client/src/Contract/Types.purs [hunk 2]
```diff
@@ -14,31 +14,41 @@ import Data.Map (Map)
...
 import Marlowe.PAB (PlutusAppId, MarloweParams)
 import Marlowe.Semantics (ChoiceId, ChosenNum, Party, Slot, TransactionInput, Accounts)
 import WalletData.Types (WalletDetails, WalletNickname)
 
+-- The Contract.State wraps the Execution.State up with some additional information that is necessary
+-- to render the execution state of the contract in the app. Some of this is derived data that could
+-- be calculated on the fly, but is stored here for efficiency. Some of it is data specific to the UX
+-- (like which step card is currently centered, and which tabs are shown in which step cards). And some
+-- of it is metadata (the actual metadata of the contract, but also functional metadata like the ID of
+-- the user's follower app for this contract, and the MarloweParams of the contract).
```
Comment (hrajchert) [inline marlowe-dashboard-client/src/Contract/Types.purs:29]: Not sure what is meant by this comment, maybe we could add the pertinent parts to the fields that need clarification, for example the fields that are stored for efficiency.

We need to keep in mind that this component changed scope since it's inception, so just `Contract` might not be the best name. Maybe it would help if we renamed this to ContractCarousel, and identify parts that we'll need to separate when we try to add the same/similar visualization to the simulator.

This is however a Component folder, so explaining 
```
Some of it is data specific to the UX (like which step card is currently centered, and which tabs are shown in which step cards)
```
is redundant.

File: marlowe-dashboard-client/src/Contract/Types.purs [hunk 3]
```diff
@@ -14,31 +14,41 @@ import Data.Map (Map)
...
-    -- (both sides inclusive). This is because the array represent the past steps and the
-    -- executionState has the current state and visually we can select any one of them.
-    , selectedStep :: Int
     , metadata :: MetaData
+    , executionState :: Execution.State
+    -- This is the tab of the current (latest) step card - previous step cards have their own tabs.
+    , tab :: Tab
+    -- These contain the data needed to render the previous step cards. This corresponds loosely to the
+    -- array of PreviousStates in the Execution.State, except that the shape and content of the data is
+    -- slightly different, and we include timeout steps here in addition to transaction steps.
```
Comment (hrajchert) [inline marlowe-dashboard-client/src/Contract/Types.purs:39]: In general, I'd try to avoid explaining the `what` if it can be "easily" understood and focus on the `why` if it's not clear.

Most of this comment focus on the `what`, which can be seen right below this data type. On the `why` we need to maintain a similar structure to the Execution.State, I'm sorry, I should have explicit it when I introduced the change.

At some point we only used to have the ExecutionState to render the past and current card, but when I implemented the timeout card, I realized that we needed to decouple the cards from the execution state. The reason is that if a card has timeouted, the actual contract does not move, it keeps its current state until you make a transaction, which can be an empty transaction just to "advance", or if the timeout has a `When` contract it can be one of the actions in the `Case` array.

We use `regenerateStepCards` to construct this array from the ExecutionState, which may include a `PendingTimeouts`. That object has a comment that kind of explain this, so we can refer to it.

File: marlowe-dashboard-client/src/Marlowe/Execution/State.purs [hunk 4]
```diff
@@ -1,133 +1,75 @@
...
-_pendingTimeouts :: Traversal' ExecutionState (Array Slot)
-_pendingTimeouts = prop (SProxy :: SProxy "mPendingTimeouts") <<< _Just <<< prop (SProxy :: SProxy "timeouts")
-
-isClosed :: ExecutionState -> Boolean
-isClosed { current: { contract: Close } } = true
-
-isClosed _ = false
-
-initExecution :: Slot -> Contract -> ExecutionState
+initExecution :: Slot -> Contract -> State
```
Comment (hrajchert) [inline marlowe-dashboard-client/src/Marlowe/Execution/State.purs:25]: between the module changes (Execution -> State, Type, Lenses) and some of the renamings, this is a little hard to follow. I'd suggest taking the feedback of this PS and separate the changes into multiple ones.

File: marlowe-dashboard-client/src/Marlowe/Execution/State.purs [hunk 5]
```diff
@@ -136,7 +78,7 @@ timeoutState currentSlot { current, previous, mPendingTimeouts, mNextTimeout } =
     advanceAllTimeouts ::
       Maybe Slot ->
       Array Slot ->
-      State ->
+      Marlowe.State ->
```
Comment (hrajchert) [inline marlowe-dashboard-client/src/Marlowe/Execution/State.purs:81]: once again, the `Marlowe` prefix does not clarify which state is this referring to

**Review by hrajchert — CHANGES_REQUESTED** [review]
I agree on separating the Execution into State, type and Lenses, but I'd do that as a separate PR.

We can further discuss the other changes

**Comment by merivale** [comment]
All fair points. I'll close this and open another PR with just the splitting of `Marlowe.Execution`, and we can discuss the rest further. In general I have two concerns:

1. Some of the names don't feel great. I agree that `marloweState` isn't great either. There are too many "states" everywhere. :confused:
2. I'm wondering about the way `Contract` and `Execution` are divided up. (This PR didn't do anything about this anyway, but it was on my mind while I was doing it.) I'm thinking we should maybe put a bit more stuff in `Execution` - specifically the `previousSteps` (minus the `tab`) and the `namedActions`. (And then perhaps the data in the `Contract.previousSteps` and the `Execution.previousState` could be merged into one array...?) Then `Contract` would just be for metadata (the actual `metadata`, as well as `marloweParams` and `followerAppId`), view state (the `tabs` and `selectedStep`), and stuff that's specific to the current user/wallet (`participants`, `userParties`).
